### PR TITLE
ref(clusters): Get Clickhouse connections from cluster

### PR DESCRIPTION
**Pull request by lynnagara**
This is a follow up to #860.

Clickhouse connections for the cleanup, optimize and replacer scripts
now come from the cluster, rather than from the soon to be deprecated
global CLICKHOUSE_HOST and CLICKHOUSE_PORT values.

The cluster now lazily creates Clickhouse connections if required,
rather than creating each possible configuration on instantiation.

Note that it is no longer possible to pass a clickhouse host without a port via the CLI

Before this PR is merged we need to:
- [x] Update cleanup script in production to pass a ClickHouse port
- [x] Update optimize script in production to pass a ClickHouse port

### CODE REVIEW

File: snuba/cli/cleanup.py
Comment (fpacifici) [inline snuba/cli/cleanup.py:64]: This script uses the local table, so before we can use the Cluster configuration we need to figure out how to obtain a connection for the local table host since this will not be the same as what we have today in settings.CLICKHOUSE_HOST. 
So if someone runs this in prod with this change without passing the CLI parameters this will fail because it will try to cleanup the wrong host.

I would not apply this change to the scripts that act on local tables before we have a final decision on how to run these in prod with the new cluster system. I believe there are still open points there or am I forgetting about something ?

File: snuba/cli/optimize.py
Comment (fpacifici) [inline snuba/cli/optimize.py:66]: Same as for cleanup. This works on the local table so on different hosts.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:15]: I think this way of defining connection parameters seems quite ad-hoc for the specific use cases we have right now. I think we may be missing a decision to make:
given that the cluster is provided in configuration and the application should move towards a place where it does not care much about the connection. parameters, should connection settings be:
- hardcoded and classified in categories so that the code picks up the connection. it needs
- in settings and classified in categories so that the code keeps identifying the connection by a logical name, that then gets resolved to parameters
- entirely up to the application, such that the application code directly provides the connection settings it needs. 
(I don't like the third) but we should agree on the direction.

Also:
Do you have already an idea about whether we can transition the replacer to use reader and writer and not be Clickhouse specific ? I think we should have at least an idea on how to do that, because this ad-hoc connection seems inconsistent with the other two.
Looking at settings, having a generic readonly and readwrite makes sense, but mixing it with an application specific definition (replacer) makes it hard to understand which kinds of   connections are supposed to be added here.

On the other hand we could move totally the opposite way and have connection identifying the use case, like query, insert, replace, optimize, etc.
Which is the direction you plan to move towards ?

File: snuba/cli/cleanup.py
Comment (fpacifici) [inline snuba/cli/cleanup.py:66]: I think this will be quite fragile.
We are running this run_cleanup on all clusters and not only on the clusters where the table is present.
This incidentally works now because run_cleanup would pick up the parts to clean by selecting the parts where table = xxx. But this is not guaranteed. 
Passing invalid arguments to a method with the assumption that it will handle them gracefully and generate correct results anyway is dangerous and easy to break.

Any way we can focus only on the cluster that handles the writer for the dataset passed in input ?

File: snuba/cli/optimize.py
Comment (fpacifici) [inline snuba/cli/optimize.py:69]: Same concern I had on cleanup. we are running optimize on all cluster when the user provided specifically one dataset. What if one of the configured cluster is down. This would prevent us from doing optimize on a separate cluster that is unrelated.

**Review by fpacifici — CHANGES_REQUESTED** [review]
I think there are a few implementation issue and we need to clarify a topic about connection settings. See comment inline

File: snuba/cli/cleanup.py
Comment (lynnagara) [reply snuba/cli/cleanup.py:64]: Yeah, this part is only intended for use in the case where we do not have distributed tables right now. Distributed must always provide the host and port. (I'm not 100% sure if this is a use case that we want to support - we can either just make host and port required removing the default, or alternatively check for `local_dataset_mode` before using the one from the relevant cluster)

Comment (fpacifici) [reply snuba/cli/cleanup.py:64]: We should try to avoid making this process different for the user between distributed and local tables, if possible at all. 
I am not familiar on how ops run this in prod. Do you have details? I think the question is whether breaking the abstraction that the cluster is fully configured through settings (and what clickhouse already knows) for this case is worth.

Comment (lynnagara) [reply snuba/cli/cleanup.py:64]: Currently, prod passes the clickhouse host on the command line but uses the default port. (the default host wouldn't work since that's the distributed one).
Single tenant does not pass any value for host or port and therefore relies on the default being applied.
If we want to keep those scripts working it seems that we have to keep the different behaviour on local and distributed mode until the cluster can become aware of all the other nodes.

File: snuba/cli/cleanup.py
Comment (lynnagara) [reply snuba/cli/cleanup.py:66]: Yes, thanks this was an oversight

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:15]: I don't know of any reasons why most this these options will need to be configurable by users (apart from the replacer_max_block_size and replacer_max_memory_usage settings that already exists). Maybe in the future being able to adjust more of these options might be needed. Right now they seem largely application dependent to me. e.g. We'll always want readonly for querying and readwrite for DDL, etc.

Unfortunately, we can't do away with the replacer specific options and have generic read and readwrite only. So moving to the second option you mentioned (connection by use case) seems the most viable option to me.

File: snuba/cli/cleanup.py
Comment (fpacifici) [inline snuba/cli/cleanup.py:57]: Have you verified we never rely on the default in any environment where we may be running in dist mode and changing the settings to connect to the specific node instead of passing the parameters in the CLI ?

File: snuba/cli/cleanup.py
Comment (fpacifici) [reply snuba/cli/cleanup.py:68]: I think this is still to broad.
we take all clusters from a dataset. Not only the writer ones. 
Then we run the cleanup for the same table on all of them, while we should be running the cleanup only on the connection for the cluster that holds the table.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:110]: I don't see this ever being used. Is this intentional ?
Also, aside from the fact that we pass this parameter during the initialization of ClickhousePool separately from the settings, is there any reason not to keep this hardcoded with the rest of the settings in ClickhouseClientSettings?
The fact that clickhouse takes settings and timeout as separate parameters is a clickhouse driver implmentation detail that should not drive our api. If we make the callsite pick the use case without specifying parameters, we should not allow the callsite to specify the timeout either.

**Review by fpacifici — COMMENTED** [review]
I think we are pretty close. Only some lower level comments

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:24]: I don't expect them to be different. I think it's a future possibility, like if we add some task that only runs on the cluster. But right now this is mostly a refactor of the existing settings which are identical for both single node and cluster.

File: snuba/cli/cleanup.py
Comment (lynnagara) [reply snuba/cli/cleanup.py:57]: We do actually pass a ClickHouse host without a port (relying on default) in production right now. I think it's worth keeping this change and making sure we update there before this PR is merged though.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py:110]: Hmm, yes. It's supposed to be an argument passed by user from the optimize CLI script (which is why it can't be provided on initialization). Fixing.

File: snuba/cli/cleanup.py
Comment (lynnagara) [reply snuba/cli/cleanup.py:68]: Ok. Cleanup only makes sense for the writable storages right? Do you know if there's any use case for running cleanup on a table that is not our writable storage table?

File: snuba/cli/cleanup.py
Comment (lynnagara) [reply snuba/cli/cleanup.py]: This is updated in prod now

File: snuba/clusters/cluster.py
Comment (fpacifici) [reply snuba/clusters/cluster.py:110]: Do we actually know the timeout is actually something we need? If we could remove it we will cut away quite a bit of complexity. I feel having this ad hoc parameter that is configured on the command line while all other connection parameter come from the cluster configuration is going against our design.
Could you please verify whether we absolutely need to maintain this field ?

**Review by fpacifici — COMMENTED** [review]
Let's just verify whether we really need to preserve the timeout CLI parameter instead of getting rid of it in favor of making it part of the cluster configuration like all other connection settings.

Otherwise it is ready.

**Comment by lynnagara** [comment]
You're right, it seems like no one was passing a different timeout value to optimize - removed the option.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:20]: I don't think the cluster (which is a low level abstraction should depend on the optimize process.  At best the opposite should be true. This seems a settings of one specific use case of clickhouse client settings (which is here). What about defining it here and make optimize depend on this instead?

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:25]: This seems a good candidate for a NamedTuple so people know what that Optional[int] is.

File: snuba/clusters/cluster.py
Comment (fpacifici) [inline snuba/clusters/cluster.py:126]: Why not 
```
settings, timeout = client_settings.value
```
?

File: snuba/optimize.py
Comment (fpacifici) [inline snuba/optimize.py:7]: See my comment on clusters

**Review by fpacifici — CHANGES_REQUESTED** [review]
two minor issues. Please ping me when they are corrected. Otherwise this is ready.

File: snuba/clusters/cluster.py
Comment (lynnagara) [reply snuba/clusters/cluster.py]: Sounds good, moved it into the client settings now.

**Review by fpacifici — APPROVED** [review]
thanks

**Review by fpacifici — APPROVED** [review]
thanks

**Pull request merged**
